sonic_package_manager/manager.py

```diff
@@ -838,10 +813,10 @@ def migrate_package(old_package_entry,
                 package_source = self.get_package_source(package_ref=new_package_ref)
                 package = package_source.get_package()
                 new_package_default_version = package.manifest['package']['version']
-                if old_package.version > new_package_default_version:
+                if old_package.version >= new_package_default_version:
```

Comment:
For instance, Third party container debian:latest with name MYDEB was downloaded to the switch( default_version will be 1.0.0 and default_reference will be latest)
Now, on attempt to migrate, the new_package_default_reference will also be 1.0.0.
In this case as well, we could migrate the MYDEB tpcm docker using image save technique of migrate_package.

Reply:
@stepanblyschak Please suggest if the below is OK.

MYDEB package (debian:latest) is installed in the current image.
Now, on sonic-sonic image upgrade, as part of new image, MYDEB package is not installed. 
Hence as part of migrate_packages(), "elif new_package.default_reference is not None:" is hit and finally the version comparison occurs where the old_package.version [1.0.0] and new_package_default_version are same [1.0.0].
As per the original code, calling self.install("MYDEB=1.0.0") in the else case would fail.

For this to work, we could change the else case code to below (if migration is not supposed to be called in version equal case)
                else:
                    #self.install(f'{new_package.name}={new_package_default_version}')
                    repo_tag_formed="{}:{}".format(new_package.repository, new_package.default_reference)
                    print("repo_tag_formed:{}".format(repo_tag_formed))
                    self.install(None,repo_tag_formed,name=new_package.name)

Reply:
@sg893052 Could you please check why you have default_reference set to 1.0.0? In your case I would expect the following will execute:
```
            else:
                # No default version and package is not installed.
                # Migrate old package same version.
                new_package.version = old_package.version
                migrate_package(old_package, new_package)
```
The ```default_reference``` is a tag or digest of Docker image that will be a default installation candidate. 1.0.0 does not exists for debian docker image, therefore you get this error.

Reply:
@stepanblyschak In the migrate_package function, the following condition cases are defined, but there is still some confusion that requires clarification:

Case 1: if new_package.installed.
Case 2: else if new_package.default_reference is not None.
Case 3: else.

In the specific example of the MYDEB package (debian:latest), where "latest" is set as the new_package.default_reference, Case 2 will be triggered because new_package.default_reference is not None. However, Case 3 will not be triggered in this scenario.

Reply:
@stepanblyschak Firstly, new_package.default_reference is "latest" which makes it to hit the case2 .

Subsequently, new_package_default_version = package.manifest['package']['version'] --> new_package_default_version is derived from package's manifest package version which is 1.0.0 .
Hence the new_package_default_version variable  is set to 1.0.0 here

Log:
==================================================================
**package**: Package(entry=PackageEntry(name='**MYDEB**', repository='debian', description='', default_reference='latest', version=None, installed=False, built_in=False, image_id=None), metadata=Metadata(**manifest**={'version': Version('1.0.0'), '**package**': {'**version**': Version('1.0.0'), 'name': 'MYDEB', 'description': '', 'base-os': ComponentConstraints(components={}), 'depends': [], 'breaks': [], 'init-cfg': {}, 'changelog': {}, 'debug-dump': ''}, 'service': {'name': 'MYDEB', 'requires': [], 'requisite': [], 'wanted-by': [], 'after': [], 'before': [], 'dependent': [], 'dependent-of': [], 'post-start-action': '', 'pre-shutdown-action': '', 'asic-service': False, 'host-service': True, 'delayed': False, 'check_up_status': False, 'warm-shutdown': {'after': [], 'before': []}, 'fast-shutdown': {'after': [], 'before': []}, 'syslog': {'support-rate-limit': False}}, 'container': {'privileged': False, 'volumes': [], 'mounts': [], 'environment': {}, 'tmpfs': []}, 'processes': [], 'cli': {'mandatory': False, 'show': [], 'config': [], 'clear': [], 'auto-generate-show': False, 'auto-generate-config': False, 'auto-generate-show-source-yang-modules': [], 'auto-generate-config-source-yang-modules': []}}, components={}, yang_modules=[]))

migrate -- old_package.version:1.0.0 new_package_default_version: 1.0.0

==================================================================

Code snippet:
```
            elif new_package.default_reference is not None:
                new_package_ref = PackageReference(new_package.name, new_package.default_reference)
                package_source = self.get_package_source(package_ref=new_package_ref)
                package = package_source.get_package()
                print("package:{}".format(package))
                new_package_default_version = package.manifest['package']['version']
                print("migrate -- old_package.version:{} new_package_default_version: {}".format(old_package.version, new_package_default_version))
                if old_package.version > new_package_default_version:

```

Reply:
@sg893052 Thanks for all the details.

I agree with your following suggestion:

> For this to work, we could change the else case code to below (if migration is not supposed to be called in version equal case)

                else:
                    #self.install(f'{new_package.name}={new_package_default_version}')
                    repo_tag_formed="{}:{}".format(new_package.repository, new_package.default_reference)
                    self.install(None,repo_tag_formed,name=new_package.name)

If the logic found `new_package.default_reference is not None` and decided to install from default_reference it should use default_reference instead of version. The code currently assumes default_reference is the same as default_version, however it's not always the case.

Reply:
Thanks!